pq-crypto/EVP_kem.c

```diff
@@ -0,0 +1,95 @@
+// -----------------------------------------------------------------------------
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// KEM Library
+// Abstract: Post-Quantum Algorithm API
+// -----------------------------------------------------------------------------
+
+#include <stdint.h>
+#include <stdlib.h>
+#include "EVP_kem.h"
+#include "internal.h"
+
+struct pq_kem evp_sike_p434_r3 = {
+        .name = "SIKEp434r3-KEM",
+        .public_key_length = SIKE_P434_R3_PUBLIC_KEY_BYTES,
+        .private_key_length = SIKE_P434_R3_PRIVATE_KEY_BYTES,
+        .shared_secret_key_length = SIKE_P434_R3_SHARED_SECRET_BYTES,
+        .ciphertext_length = SIKE_P434_R3_CIPHERTEXT_BYTES,
+        .generate_keypair = &sike_p434_r3_crypto_kem_keypair,
+        .encapsulate = &sike_p434_r3_crypto_kem_enc,
+        .decapsulate = &sike_p434_r3_crypto_kem_dec,
+};
+
+int pq_kem_params_alloc(pq_kem kem, pq_kem_params *key_params) {
+    key_params->kem = &evp_sike_p434_r3;
+    key_params->public_key = malloc(key_params->kem->public_key_length);
+    key_params->private_key = malloc(key_params->kem->private_key_length);
+    key_params->ciphertext = malloc(key_params->kem->ciphertext_length);
+    key_params->shared_secret = malloc(key_params->kem->shared_secret_key_length);
+    // SUCCESS
+    return 1;
+}
+
+int pq_kem_params_free(pq_kem_params *key_params) {
+    if (key_params != NULL) {
+        free(key_params->public_key);
+        free(key_params->private_key);
+        free(key_params->ciphertext);
+        free(key_params->shared_secret);
```

Comment:
If we want to be really safe here, we should check all of the nested pointers for NULL before attempting to free them e.g.

`if (key_params->public_key != NULL) {free(key_params->public_key;}`
...